In [71]:
import pandas as pd # A general purpose Python library for data analysis
import numpy as np # A library for scientific computing in Python (e.g., provides high-performance multi-dimensional array objects and operations)

import matplotlib.pyplot as plt # a plotting library for Python and NumPy (readily customizable)
import seaborn as sns # Another plotting library for Python (fewer syntax, excellent default themes, behind the scenes, it uses matplotlib)
import time

## Knowledge Stream Summer 2023

In this notebook, we will learn about the key data structures provided by the Pandas library: **Data Frames, Series, and Indices**.

In addition, we will learn about the following operations:
* How to access data contained in these structures?
* How to read files (e.g., csv, xlsx, sql) to create these structures?
* How to carry out different data manipulation tasks using these structures?

`Dataset`: US elections with information about candidates, their party, votes won, year of election and the result.

## Reading in Data Frames from Files

Pandas has a number of useful file reading tools. You can see them enumerated by typing **"pd.re"** and pressing `tab`. We'll be using **read_csv** today. Note that these file reading functions do all the *data parsing* for you, which is very useful.

Before loading a file into a dataframe, let's first take a look at the **elections.csv** file

In [96]:
#Answer Here
d = pd.read_csv("H:\Machine Learning\Excel files\elections.csv")

copy_df = d.copy()

We can use the **head command** to show only a few rows of a dataframe.

# heading
## heading2

In [73]:
# Answer Here
d.head(1)


Candidate       Party     %  Year Result
0    Reagan  Republican  50.7  1980    win

There is also a **tail command**.

In [74]:
#Answer Here
d.tail(3)

Candidate       Party     %  Year Result
20    Romney  Republican  47.2  2012   loss
21   Clinton  Democratic  48.2  2016   loss
22     Trump  Republican  46.1  2016    win

The `read_csv` command lets us specify a **column to use an index**. For example, we could have used __Year__ as the index.

In [75]:
#Answer Here 
d = pd.read_csv("H:\Machine Learning\Excel files\elections.csv",index_col="Year")

Alternately, we could have used the **set_index** commmand on the dataframe.

In [76]:
d.reset_index(inplace=True)

In [77]:
#Answer Here
d.set_index("Party",inplace=True)


# Caution:
The **set_index command** (along with all other data frame methods) **does not modify the dataframe**, i.e., the original "elections" is untouched. Note: There is a flag called "inplace" which does modify the calling dataframe (e.g., `elections.set_index("Party",inplace=True)`).

In [78]:
d.reset_index(inplace=True)

## Duplicate Columns?
By contast, column names MUST be unique. For example, if we try to read in a file for which column names are not unique, Pandas will automatically any duplicates. Load duplicate_columns.csv

In [79]:
#Answer Here
dup_cols = pd.read_csv("H:\Machine Learning\Excel files\duplicate_columns.csv")
dup_cols

name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana

In [80]:
print(d)

          Party  Year Candidate     % Result
0    Republican  1980    Reagan  50.7    win
1    Democratic  1980    Carter  41.0   loss
2   Independent  1980  Anderson   6.6   loss
3    Republican  1984    Reagan  58.8    win
4    Democratic  1984   Mondale  37.6   loss
5    Republican  1988      Bush  53.4    win
6    Democratic  1988   Dukakis  45.6   loss
7    Democratic  1992   Clinton  43.0    win
8    Republican  1992      Bush  37.4   loss
9   Independent  1992     Perot  18.9   loss
10   Democratic  1996   Clinton  49.2    win
11   Republican  1996      Dole  40.7   loss
12  Independent  1996     Perot   8.4   loss
13   Democratic  2000      Gore  48.4   loss
14   Republican  2000      Bush  47.9    win
15   Democratic  2004     Kerry  48.3   loss
16   Republican  2004      Bush  50.7    win
17   Democratic  2008     Obama  52.9    win
18   Republican  2008    McCain  45.7   loss
19   Democratic  2012     Obama  51.1    win
20   Republican  2012    Romney  47.2   loss
21   Democ

## The [ ] Operator & Indexing

The DataFrame class has an indexing operator **[ ]** (also known as the 'brack' operator) that lets you do a variety of different things. If your provide a String to the **[ ]** operator, you get back a ***Series*** corresponding to the requested label.

1.Use **[ ]** to display different columns

2.Use List retrive multiple columns

In [81]:
# Answer Here
# 1- Display one column
d["Candidate"]

print(d)

          Party  Year Candidate     % Result
0    Republican  1980    Reagan  50.7    win
1    Democratic  1980    Carter  41.0   loss
2   Independent  1980  Anderson   6.6   loss
3    Republican  1984    Reagan  58.8    win
4    Democratic  1984   Mondale  37.6   loss
5    Republican  1988      Bush  53.4    win
6    Democratic  1988   Dukakis  45.6   loss
7    Democratic  1992   Clinton  43.0    win
8    Republican  1992      Bush  37.4   loss
9   Independent  1992     Perot  18.9   loss
10   Democratic  1996   Clinton  49.2    win
11   Republican  1996      Dole  40.7   loss
12  Independent  1996     Perot   8.4   loss
13   Democratic  2000      Gore  48.4   loss
14   Republican  2000      Bush  47.9    win
15   Democratic  2004     Kerry  48.3   loss
16   Republican  2004      Bush  50.7    win
17   Democratic  2008     Obama  52.9    win
18   Republican  2008    McCain  45.7   loss
19   Democratic  2012     Obama  51.1    win
20   Republican  2012    Romney  47.2   loss
21   Democ

In [82]:
d.reset_index(inplace=True)

The **[ ]** operator also accepts a list of strings. In this case, you get back a **DataFrame** corresponding to the requested strings.

In [83]:
# Answer Here
#2-Display Columns using list
d[["Candidate","Party"]]

Candidate        Party
0     Reagan   Republican
1     Carter   Democratic
2   Anderson  Independent
3     Reagan   Republican
4    Mondale   Democratic
5       Bush   Republican
6    Dukakis   Democratic
7    Clinton   Democratic
8       Bush   Republican
9      Perot  Independent
10   Clinton   Democratic
11      Dole   Republican
12     Perot  Independent
13      Gore   Democratic
14      Bush   Republican
15     Kerry   Democratic
16      Bush   Republican
17     Obama   Democratic
18    McCain   Republican
19     Obama   Democratic
20    Romney   Republican
21   Clinton   Democratic
22     Trump   Republican

A list of one label also returns a DataFrame. This can be handy if you want your results as a DataFrame, not a series.

Note that we can also use the **to_frame** method to turn a Series into a DataFrame.

Extract one col name "Candidates" from DataFrame it will be a series. Convert series into a DataFrame.

In [84]:
# Answer Here
#Extract Candidate column
d.Candidate


0       Reagan
1       Carter
2     Anderson
3       Reagan
4      Mondale
5         Bush
6      Dukakis
7      Clinton
8         Bush
9        Perot
10     Clinton
11        Dole
12       Perot
13        Gore
14        Bush
15       Kerry
16        Bush
17       Obama
18      McCain
19       Obama
20      Romney
21     Clinton
22       Trump
Name: Candidate, dtype: object

In [88]:
# Answer Here
#convert candidate series to data frame
a = d.Candidate.to_frame()
print(a)

   Candidate
0     Reagan
1     Carter
2   Anderson
3     Reagan
4    Mondale
5       Bush
6    Dukakis
7    Clinton
8       Bush
9      Perot
10   Clinton
11      Dole
12     Perot
13      Gore
14      Bush
15     Kerry
16      Bush
17     Obama
18    McCain
19     Obama
20    Romney
21   Clinton
22     Trump


### Row Indexing

The `[]` operator also accepts numerical slices as arguments. In this case, we are indexing by row, not column!

Extract few rows from DataFrame

In [89]:
# Answer Here
#Extract rows from data frame

d[0:6]



Candidate        Party     %  Year Result
0    Reagan   Republican  50.7  1980    win
1    Carter   Democratic  41.0  1980   loss
2  Anderson  Independent   6.6  1980   loss
3    Reagan   Republican  58.8  1984    win
4   Mondale   Democratic  37.6  1984   loss
5      Bush   Republican  53.4  1988    win

If you provide a single argument to the `[]` operator, it tries to use it as a name. This is true even if the argument passed to **[ ]** is an integer.

In [90]:
#elections[0] #this does not work, try uncommenting this to see it fail in action, woo


#d[0]   

 # As mentioned , This will through an error , uncomment to check its error 

The following cells allow you to **test your understanding**. Let's go over the summary of what we have learnt (see slides).

# Creating DataFrames
Create DataFrame using List and Columns name.

In [91]:
# Answer Here

data_frame = pd.DataFrame([1,2,3,4],columns=["First Column"])
data_frame

First Column
0             1
1             2
2             3
3             4

Creating DataFrames using **Dictionary**.

In [92]:
# Answer Here

di = { "Key1" : [1,2,3,4],
      "key2"  : ["one","two","three","four"]
}

dict_to_df = pd.DataFrame(di)
dict_to_df

Key1   key2
0     1    one
1     2    two
2     3  three
3     4   four

## Filtering via Boolean Array Selection

The `[]` operator also supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows. The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

In [93]:
d[[False, False, False, False, False,
          False, False, True, False, False,
          True, False, False, False, True,
          False, False, False, False, False,
          False, True, False]]

    

Candidate       Party     %  Year Result
7    Clinton  Democratic  43.0  1992    win
10   Clinton  Democratic  49.2  1996    win
14      Bush  Republican  47.9  2000    win
21   Clinton  Democratic  48.2  2016   loss

In [95]:
d[d["Result"]=="win"]

Candidate       Party     %  Year Result
0     Reagan  Republican  50.7  1980    win
3     Reagan  Republican  58.8  1984    win
5       Bush  Republican  53.4  1988    win
7    Clinton  Democratic  43.0  1992    win
10   Clinton  Democratic  49.2  1996    win
14      Bush  Republican  47.9  2000    win
16      Bush  Republican  50.7  2004    win
17     Obama  Democratic  52.9  2008    win
19     Obama  Democratic  51.1  2012    win
22     Trump  Republican  46.1  2016    win

One very common task in Data Science is **filtering**. Boolean Array Selection is one way to achieve this in Pandas. We start by observing that **logical operators** like the equality operator can be applied to **Pandas Series data** to generate a **Boolean Array**.

Compare the 'Result' column to the String 'win' and Show results

In [97]:
#Answer Here
copy_df[copy_df["Result"] == "win"]

Candidate       Party     %  Year Result
0     Reagan  Republican  50.7  1980    win
3     Reagan  Republican  58.8  1984    win
5       Bush  Republican  53.4  1988    win
7    Clinton  Democratic  43.0  1992    win
10   Clinton  Democratic  49.2  1996    win
14      Bush  Republican  47.9  2000    win
16      Bush  Republican  50.7  2004    win
17     Obama  Democratic  52.9  2008    win
19     Obama  Democratic  51.1  2012    win
22     Trump  Republican  46.1  2016    win

Compare the 'Party' column to the String 'Democratic' and Show results

In [98]:
#Answer Here
d[d["Party"] == "Democratic"]

Candidate       Party     %  Year Result
1     Carter  Democratic  41.0  1980   loss
4    Mondale  Democratic  37.6  1984   loss
6    Dukakis  Democratic  45.6  1988   loss
7    Clinton  Democratic  43.0  1992    win
10   Clinton  Democratic  49.2  1996    win
13      Gore  Democratic  48.4  2000   loss
15     Kerry  Democratic  48.3  2004   loss
17     Obama  Democratic  52.9  2008    win
19     Obama  Democratic  51.1  2012    win
21   Clinton  Democratic  48.2  2016   loss

The output of the logical operator applied to the Series is **another Series with the same name and index, but of datatype boolean**.

These boolean Series can be used as an argument to the `[]` operator.

Creates  DataFrame of all election winners since 1980.

In [99]:
#Answer Here


a = d[d["Result"] == "win"]

election_win = pd.DataFrame(a)

election_win

Candidate       Party     %  Year Result
0     Reagan  Republican  50.7  1980    win
3     Reagan  Republican  58.8  1984    win
5       Bush  Republican  53.4  1988    win
7    Clinton  Democratic  43.0  1992    win
10   Clinton  Democratic  49.2  1996    win
14      Bush  Republican  47.9  2000    win
16      Bush  Republican  50.7  2004    win
17     Obama  Democratic  52.9  2008    win
19     Obama  Democratic  51.1  2012    win
22     Trump  Republican  46.1  2016    win

Above, we've assigned the result of the logical operator to a new variable called `iswin`. This is uncommon. Usually, the series is created and used on the same line. Such code is a little tricky to read at first, but you'll get used to it quickly.

Show all 'win' results between 1980 to 2000

In [100]:
#Answer Here
#this will pick the win results in the year range between 1980 to 2000


election_win[(election_win["Year"] >= 1980)  & ( election_win["Year"] <= 2000)]

Candidate       Party     %  Year Result
0     Reagan  Republican  50.7  1980    win
3     Reagan  Republican  58.8  1984    win
5       Bush  Republican  53.4  1988    win
7    Clinton  Democratic  43.0  1992    win
10   Clinton  Democratic  49.2  1996    win
14      Bush  Republican  47.9  2000    win

Show all 'Loss' results of Independent party

In [101]:
# Answer Here
ind_loss = d[(d["Result"] == "loss") & (d["Party"] == "Independent")]
ind_loss 

Candidate        Party     %  Year Result
2   Anderson  Independent   6.6  1980   loss
9      Perot  Independent  18.9  1992   loss
12     Perot  Independent   8.4  1996   loss

We can select multiple criteria by creating multiple boolean Series and combining them using the `&` operator.

Show results of win with percentage less than 50%

In [102]:
# Answer Here
dresults = d[(d["Result"]=="win") & (d["%"] < 50)]
dresults 

Candidate       Party     %  Year Result
7    Clinton  Democratic  43.0  1992    win
10   Clinton  Democratic  49.2  1996    win
14      Bush  Republican  47.9  2000    win
22     Trump  Republican  46.1  2016    win

Show all 'win' results between 1980 to 2000

In [103]:
# Answer Here
win_results = d[d["Result"] == "win"] 
win_results[(win_results["Year"] >= 1980) & (win_results["Year"]<= 2000) ]

Candidate       Party     %  Year Result
0     Reagan  Republican  50.7  1980    win
3     Reagan  Republican  58.8  1984    win
5       Bush  Republican  53.4  1988    win
7    Clinton  Democratic  43.0  1992    win
10   Clinton  Democratic  49.2  1996    win
14      Bush  Republican  47.9  2000    win

## Loc and iLoc

Show 5 enteries from start

In [104]:
# Answer Here
d.iloc[0:5]

Candidate        Party     %  Year Result
0    Reagan   Republican  50.7  1980    win
1    Carter   Democratic  41.0  1980   loss
2  Anderson  Independent   6.6  1980   loss
3    Reagan   Republican  58.8  1984    win
4   Mondale   Democratic  37.6  1984   loss

You can provide `.loc` a list of row labels [0-5] and column labels ['Candidate','Party', 'Year'] as input to return a dataframe

In [105]:
#Answer Here

d.loc[0:5,['Candidate','Party', 'Year']]

Candidate        Party  Year
0    Reagan   Republican  1980
1    Carter   Democratic  1980
2  Anderson  Independent  1980
3    Reagan   Republican  1984
4   Mondale   Democratic  1984
5      Bush   Republican  1988

Loc also supports **slicing** (for all types, including numeric and string labels!). Note that the slicing for loc is **inclusive**, even for numeric slices.

Use Slicing on Rows and Columns

In [106]:
# Answer Here

d.loc[0:3,"Candidate":"%"]

Candidate        Party     %
0    Reagan   Republican  50.7
1    Carter   Democratic  41.0
2  Anderson  Independent   6.6
3    Reagan   Republican  58.8

If we provide only a **single label** for the column argument, we get back a **Series**.

In [107]:
# Answer Here
#2 ways to do this

#1- using []
d["Candidate"]

#2- using .
d.Candidate

0       Reagan
1       Carter
2     Anderson
3       Reagan
4      Mondale
5         Bush
6      Dukakis
7      Clinton
8         Bush
9        Perot
10     Clinton
11        Dole
12       Perot
13        Gore
14        Bush
15       Kerry
16        Bush
17       Obama
18      McCain
19       Obama
20      Romney
21     Clinton
22       Trump
Name: Candidate, dtype: object

If we give only one row but many column labels, we'll get back a **Series** corresponding to a row of the table. This new Series has a neat index, where **each entry is the name of the column** that the data came from.

In [108]:
# Answer Here
d.loc[0,["Candidate","Party"]]

Candidate        Reagan
Party        Republican
Name: 0, dtype: object

If we omit the column argument altogether, the **default behavior is to retrieve all columns**.

In [109]:
# Answer Here
d.loc[0]

Candidate        Reagan
Party        Republican
%                  50.7
Year               1980
Result              win
Name: 0, dtype: object

Specify Rows and Columns as List to retrive specific enteries

In [110]:
# Answer Here

d.loc[[0,2,5],["Party","Candidate"]]

Party Candidate
0   Republican    Reagan
2  Independent  Anderson
5   Republican      Bush

## String-labeled Rows

Let's do a quick example using data with string-labeled rows instead of integer labeled rows, just to make sure we're really understanding loc.

Use mottos.csv file

In [112]:
# Answer Here
data2= pd.read_csv("Excel files\mottos.csv")
data2.head(3)

State                          Motto                 Translation  \
0  Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1   Alaska            North to the future                           —   
2  Arizona                     Ditat Deus                God enriches   

  Language Date Adopted  
0    Latin         1923  
1  English         1967  
2    Latin         1863

Extract slice, can be specified using slice notation, even if the rows have string labels instead of integer labels.

In [113]:
data2[0:5]

State                          Motto                 Translation  \
0     Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1      Alaska            North to the future                           —   
2     Arizona                     Ditat Deus                God enriches   
3    Arkansas                 Regnat populus             The people rule   
4  California                Eureka (Εὕρηκα)             I have found it   

  Language Date Adopted  
0    Latin         1923  
1  English         1967  
2    Latin         1863  
3    Latin         1907  
4    Greek         1849

### iloc

loc's cousin iloc is very similar, but is used to access based on numerical position instead of label. For example, to access to the top 3 rows and top 3 columns of a table, we can use [0:3, 0:3]. 'iloc' slicing is **exclusive**, just like standard Python slicing of numerical values.

Use iloc to extract first 3 rows and columns from elections DataFrame

In [114]:
#Answer Here
d.iloc[0:3,0:3]

Candidate        Party     %
0    Reagan   Republican  50.7
1    Carter   Democratic  41.0
2  Anderson  Independent   6.6

We will use both `loc` and `iloc` in the course. `loc` is generally preferred for a number of reasons, for example:

1. It is harder to make mistakes since you have to literally write out what you want to get.
2. Code is easier to read, because the reader doesn't have to know e.g., what column #17 represents.
3. It is robust against permutations of the data, e.g. the social security administration switches the order of two columns.

However, iloc is sometimes more convenient. We'll provide examples of when iloc is the superior choice.

## Handy Properties and Utility Functions for Series and DataFrames

The head, shape, size, and describe methods can be used to quickly get a good sense of the data we're working with. For example:

In [ ]:
# Answer Here
data2.head()

State                          Motto                 Translation  \
0     Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1      Alaska            North to the future                           —   
2     Arizona                     Ditat Deus                God enriches   
3    Arkansas                 Regnat populus             The people rule   
4  California                Eureka (Εὕρηκα)             I have found it   

  Language Date Adopted  
0    Latin         1923  
1  English         1967  
2    Latin         1863  
3    Latin         1907  
4    Greek         1849

Size of DataFrame

In [115]:
# Answer Here
data2.size

250

The fact that the size is 250 means our data file is relatively small, with only 250 total entries.

Shape of DataFrame

In [116]:
# Answer Here
data2.shape

(50, 5)

Use describe function and extract the meaningful information from DataFrame

In [117]:
# Answer Here
data2.describe()

State                          Motto Translation Language  \
count        50                             50          49       50   
unique       50                             50          30        8   
top     Alabama  Audemus jura nostra defendere           —    Latin   
freq          1                              1          20       23   

       Date Adopted  
count            50  
unique           47  
top            1893  
freq              2

Above, we see a quick summary of all the data. For example, the most common language for mottos is Latin, which covers 23 different states. Does anything else seem surprising?

We can get a direct reference to the index using .index.

In [118]:
# Answer Here
data2.index

RangeIndex(start=0, stop=50, step=1)

It turns out the columns also have an Index. We can access this index by using `.columns`.

In [120]:
# Answer Here
data2.columns

Index(['State', 'Motto', 'Translation', 'Language', 'Date Adopted'], dtype='object')

## Sorting and Value Counts

There are also a ton of useful utility methods we can use with Data Frames and Series. For example, we can create a copy of a data frame sorted by a specific column using `sort_values`.

In [125]:
# Answer Here
data2.sort_values(by="Date Adopted")

State                                              Motto  \
20   Massachusetts          Ense petit placidam sub libertate quietem   
45        Virginia                                Sic semper tyrannis   
31        New York                                          Excelsior   
9          Georgia                        Wisdom, Justice, Moderation   
12        Illinois                  State sovereignty, national union   
18           Maine                                             Dirigo   
14            Iowa  Our liberties we prize and our rights we will ...   
7         Delaware                           Liberty and Independence   
4       California                                    Eureka (Εὕρηκα)   
48       Wisconsin                                            Forward   
36          Oregon                                Alis volat propriis   
15          Kansas                                Ad astra per aspera   
22       Minnesota                                   L'étoile du Nord   
2          Arizona                                         Ditat Deus   
26        Nebraska                            Equality before the law   
8          Florida                                    In God We Trust   
19        Maryland                       Fatti maschii, parole femine   
37    Pennsylvania                  Virtue, liberty, and independence   
40    South Dakota                          Under God the people rule   
30      New Mexico                                      Crescit eundo   
11           Idaho                                      Esto perpetua   
32  North Carolina                                   Esse quam videri   
49         Wyoming                                       Equal Rights   
17       Louisiana                         Union, justice, confidence   
3         Arkansas                                     Regnat populus   
0          Alabama                      Audemus jura nostra defendere   
42           Texas                                         Friendship   
13         Indiana                          The Crossroads of America   
28   New Hampshire                                   Live Free or Die   
1           Alaska                                North to the future   
16        Kentucky                               Deo gratiam habeamus   
44         Vermont                       Stella quarta decima fulgeat   
27          Nevada                                All For Our Country   
23     Mississippi                                   Virtute et armis   
25         Montana                                        Oro y plata   
24        Missouri                      Salus populi suprema lex esto   
10          Hawaii                  Ua mau ke ea o ka ʻāina i ka pono   
21        Michigan          Si quaeris peninsulam amoenam circumspice   
35        Oklahoma                                 Labor omnia vincit   
33    North Dakota                    \nSerit ut alteri saeclo prosit   
29      New Jersey                             Liberty and prosperity   
39  South Carolina          Dum spiro spero \nAnimis opibusque parati   
41       Tennessee                           Agriculture and Commerce   
43            Utah                                           Industry   
38    Rhode Island                                               Hope   
5         Colorado                                    Nil sine numine   
34            Ohio                  With God, all things are possible   
6      Connecticut                            Qui transtulit sustinet   
47   West Virginia                              Montani semper liberi   
46      Washington                                      Al-ki or Alki   

                                          Translation        Language  \
20  By the sword we seek peace, but peace only und...           Latin   
45                             Thus always to tyrants           Latin   
31                                        Ever upward           Latin 

As mentioned before, all Data Frame methods return a copy and do **not** modify the original data structure, unless you set inplace to True.

If we want to sort in reverse order, we can set `ascending=False`.

In [126]:
d.sort_values('%', ascending=False)

Candidate        Party     %  Year Result
3     Reagan   Republican  58.8  1984    win
5       Bush   Republican  53.4  1988    win
17     Obama   Democratic  52.9  2008    win
19     Obama   Democratic  51.1  2012    win
0     Reagan   Republican  50.7  1980    win
16      Bush   Republican  50.7  2004    win
10   Clinton   Democratic  49.2  1996    win
13      Gore   Democratic  48.4  2000   loss
15     Kerry   Democratic  48.3  2004   loss
21   Clinton   Democratic  48.2  2016   loss
14      Bush   Republican  47.9  2000    win
20    Romney   Republican  47.2  2012   loss
22     Trump   Republican  46.1  2016    win
18    McCain   Republican  45.7  2008   loss
6    Dukakis   Democratic  45.6  1988   loss
7    Clinton   Democratic  43.0  1992    win
1     Carter   Democratic  41.0  1980   loss
11      Dole   Republican  40.7  1996   loss
4    Mondale   Democratic  37.6  1984   loss
8       Bush   Republican  37.4  1992   loss
9      Perot  Independent  18.9  1992   loss
12     Perot  Independent   8.4  1996   loss
2   Anderson  Independent   6.6  1980   loss

We can also use `sort_values` on Series objects.

In [127]:
data2['Language'].sort_values().head(20)

46    Chinook Jargon
49           English
29           English
28           English
27           English
26           English
48           English
37           English
38           English
40           English
17           English
34           English
42           English
14           English
41           English
12           English
1            English
13           English
8            English
7            English
Name: Language, dtype: object

For Series, the `value_counts` method is often quite handy.

In [128]:
data2['Language'].value_counts()

Language
Latin             23
English           21
Greek              1
Hawaiian           1
Italian            1
French             1
Spanish            1
Chinook Jargon     1
Name: count, dtype: int64

Also commonly used is the `unique` method, which returns **all unique values** as a numpy array.

In [129]:
data2['Language'].unique()

array(['Latin', 'English', 'Greek', 'Hawaiian', 'Italian', 'French',
       'Spanish', 'Chinook Jargon'], dtype=object)

In [130]:
def fiba(n):
    if n < 2:
        return n
    else:
        return fiba(n-1) + fiba(n-2)



fiba(3)

2

# Thank you!